File: src/code/linked-list/DoublyLinkedList.java
class DListNode {
    int val;
    DListNode prev, next;
    DListNode(int val) { this.val = val; }
}

class DoublyLinkedList {
    DListNode head, tail;
    int size;

    void addHead(int val) {                 // 头部插入 O(1)
        DListNode node = new DListNode(val);
        if (head == null) {
            head = tail = node;
        } else {
            node.next = head;
            head.prev = node;
            head = node;
        }
        size++;
    }

    void addTail(int val) {                 // 尾部插入 O(1)
        DListNode node = new DListNode(val);
        if (tail == null) {
            head = tail = node;
        } else {
            node.prev = tail;
            tail.next = node;
            tail = node;
        }
        size++;
    }

    void insert(int index, int val) {       // 在第 index 位插入 O(n)
        if (index <= 0) { addHead(val); return; }
        if (index >= size) { addTail(val); return; }
        DListNode cur = head;
        for (int i = 0; i < index; i++) cur = cur.next;
        DListNode node = new DListNode(val);
        node.prev = cur.prev;
        node.next = cur;
        cur.prev.next = node;
        cur.prev = node;
        size++;
    }

    void remove(int val) {                  // 删除首个等于 val 的节点 O(n)
        DListNode cur = head;
        while (cur != null) {
            if (cur.val == val) {
                if (cur.prev != null) cur.prev.next = cur.next;
                else head = cur.next;
                if (cur.next != null) cur.next.prev = cur.prev;
                else tail = cur.prev;
                size--;
                break;
            }
            cur = cur.next;
        }
    }
}

File: src/code/linked-list/doubly_linked_list.go
package linkedlist

type DListNode struct {
	Val        int
	Prev, Next *DListNode
}

type DoublyLinkedList struct {
	Head, Tail *DListNode
	Size       int
}

func (l *DoublyLinkedList) AddHead(val int) { // 头部插入 O(1)
	node := &DListNode{Val: val}
	if l.Head == nil {
		l.Head, l.Tail = node, node
	} else {
		node.Next = l.Head
		l.Head.Prev = node
		l.Head = node
	}
	l.Size++
}

func (l *DoublyLinkedList) AddTail(val int) { // 尾部插入 O(1)
	node := &DListNode{Val: val}
	if l.Tail == nil {
		l.Head, l.Tail = node, node
	} else {
		node.Prev = l.Tail
		l.Tail.Next = node
		l.Tail = node
	}
	l.Size++
}

func (l *DoublyLinkedList) Insert(index, val int) { // 在第 index 位插入 O(n)
	if index <= 0 {
		l.AddHead(val)
		return
	}
	if index >= l.Size {
		l.AddTail(val)
		return
	}
	cur := l.Head
	for i := 0; i < index; i++ {
		cur = cur.Next
	}
	node := &DListNode{Val: val, Prev: cur.Prev, Next: cur}
	cur.Prev.Next = node
	cur.Prev = node
	l.Size++
}

func (l *DoublyLinkedList) Remove(val int) { // 删除首个等于 val 的节点 O(n)
	cur := l.Head
	for cur != nil {
		if cur.Val == val {
			if cur.Prev != nil {
				cur.Prev.Next = cur.Next
			} else {
				l.Head = cur.Next
			}
			if cur.Next != nil {
				cur.Next.Prev = cur.Prev
			} else {
				l.Tail = cur.Prev
			}
			l.Size--
			break
		}
		cur = cur.Next
	}
}
